### PR TITLE
Submit homework week 2 - Oguzhan

### PR DESCRIPTION
**Pull request by oguzhanturgut**
**Part 1**

* `homework.js `

**Part 2**

* `language_trigger.md`  Trigger for `countrylanguage` table and testing

**Part 3**

* `Todo-ERD.png` Todo schema ERD image export
* `todo_erd.mwb` Todo schema ERD Mysql Workbench source file

### CODE REVIEW

File: Week2/homework.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      country,
+      city,
+      countrylanguage
+    WHERE
+      country.code = city.countrycode
+        AND country.code = countrylanguage.countrycode
+        AND countrylanguage.language = ?
+    GROUP BY language;
+    `,
+    language,
```
Comment (mogaleaf) [inline Week2/homework.js]: It's working but you can make it a lot easier, do you need to join on the three tables ?

File: Week2/homework.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    `
+    SELECT DISTINCT
+      countrylanguage.language
+    FROM
+      country,
+      countrylanguage
+    WHERE
+      countrylanguage.countrycode = country.code
+        AND country.region = ?
+    ORDER BY language`,
```
Comment (mogaleaf) [inline Week2/homework.js]: Ok it works

File: Week2/homework.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  const result = await runQuery(
+    `
+    SELECT 
+      city.name AS capital
+    FROM
+      country,
+      city
+    WHERE
+      country.capital = city.id
+        AND country.name = ?`,
```
Comment (mogaleaf) [inline Week2/homework.js]: ok !

File: Week2/homework.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    COUNT(countrylanguage.language) AS numberOfLanguages
+    FROM
+      country,
+      countrylanguage
+    WHERE
+      country.code = countrylanguage.countrycode
+    GROUP BY continent
+    ORDER BY numberOfLanguages DESC`,
+  );
+
```
Comment (mogaleaf) [inline Week2/homework.js]: What is english is spoken by more than one country in america ?

File: Week2/language_trigger.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                DECLARE numberOfLangs int ;
+                SELECT count(language) INTO numberOfLangs FROM countrylanguage WHERE countrycode=new.countrycode;
+                IF numberOfLangs = 9
+                THEN
+                    SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Maximum number of language for a country (10) exceeded';
+                END IF;
+            END //
+
+delimiter ;
+```
```
[the remark this answers is not in the document]
Comment (oguzhanturgut) [reply Week2/language_trigger.md]: I missed that point the logical check should be `IF numberOfLangs > 9`

[the remark this answers is not in the document]
Comment (oguzhanturgut) [reply Week2/language_trigger.md]: I missed that point the logical check should be `IF numberOfLangs > 9`

File: Week2/homework.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    ORDER BY country`,
+    [regionName, language],
+  );
+  console.log(
+    result.length
+      ? `The ${language} spoken countries in ${regionName} : ${result
+          .map(c => c.country)
+          .toString()}`
+      : 'FALSE',
+  );
```
[the remark this answers is not in the document]
Comment (oguzhanturgut) [reply Week2/homework.js]: This query works as wanted in the requirements but i don't think this is the best practice. But after some researching i see that Mysql (but MariaDB has) don't have `with` clause for refactoring this spaghetti query. Besides i couldn't find any good solution like `NVL` in Oracle. I am really curious about your solution 🙄

```
(SELECT 
    country.name country
FROM
    country,
    countrylanguage
WHERE
    countrylanguage.countrycode = country.code
        AND country.region = 'North America'
        AND countrylanguage.language = 'English'
ORDER BY country) UNION SELECT 
    'FALSE'
FROM DUAL WHERE
    NOT EXISTS( SELECT 
            country.name country
        FROM
            country,
            countrylanguage
        WHERE
            countrylanguage.countrycode = country.code
                AND country.region = 'North America'
                AND countrylanguage.language = 'English'
        ORDER BY country);
```

File: Week2/homework.js (hunk 4 above)
Comment (oguzhanturgut) [reply Week2/homework.js]: I missed repeating languages for a continent the query must be like with `distinct countrylanguage.language`

```
SELECT 
      country.continent,
    COUNT(distinct countrylanguage.language) AS numberOfLanguages
    FROM
      country,
      countrylanguage
    WHERE
      country.code = countrylanguage.countrycode
    GROUP BY continent
    ORDER BY numberOfLanguages DESC
```

File: Week2/homework.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (mogaleaf) [reply Week2/homework.js]: Mine is almost the same . I don't say it's best to do it all in sql , I actually prefer the coding solution, it's only for the challenge.
`
SELECT DISTINCT
	country.name country
FROM
	country,
	countrylanguage
WHERE
	countrylanguage.countrycode = country.code
	AND country.region = 'Western Europe'
	AND countrylanguage.language = 'french'

UNION ALL

SELECT
	'FALSE' country
WHERE
	NOT EXISTS (
		SELECT
			1
		FROM
			country,
			countrylanguage
		WHERE
			countrylanguage.countrycode = country.code
			AND country.region = 'Western Europe'
			AND countrylanguage.language = 'french')
			
order by country;
`

File: Week2/homework.js (hunk 1 above)
Comment (oguzhanturgut) [reply Week2/homework.js]: I tried with subquery but still using three tables. and become more complex. Am i missing something? As for the references between three tables i couldn't see any solution without using all of them.

```
SELECT 
    language, SUM(result.cnt)
FROM
    countrylanguage,
    (SELECT 
        country.code countryCode, COUNT(city.name) cnt
    FROM
        country, city
    WHERE
        country.Code = city.CountryCode
    GROUP BY countryCode) result
WHERE
    result.countrycode = countrylanguage.CountryCode
GROUP BY language
ORDER BY 2 DESC;
```

File: Week2/homework.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (oguzhanturgut) [reply Week2/homework.js]: That was a very educative challenge. Thanks 🙏

**Comment by oguzhanturgut** [comment]
Thanks @mogaleaf for your time, interest and challenging comments on homework. That was really pleasure for me

File: Week2/homework.js (hunk 1 above)
Comment (mogaleaf) [reply Week2/homework.js]: What we want here is the number of city speaking a language right , So it doesn't involved Country . You can only use city and countrylanguage, joining on countrycode (countrycode is in city AND countrylanguage) where the language is <?>

Comment (oguzhanturgut) [reply Week2/homework.js]: You are totally right. Thanks for the tip. I missed the point that city and countrylanguage tables have both countrycode column. This query is the easier one with two tables: 

```
SELECT 
      countrylanguage.language, COUNT(city.name) numberOfCities
    FROM
      city,
      countrylanguage
    WHERE city.countrycode = countrylanguage.countrycode
        AND countrylanguage.language = 'English'
    GROUP BY language;
```

Comment (mogaleaf) [reply Week2/homework.js]: Almost, you can do even simpler, no need of the group by here :)

Comment (oguzhanturgut) [reply Week2/homework.js]: Making a group by over a column which is in the where clause and also selecting it... That sounds meaningless 😀
Hope this is all done now: 
```
SELECT 
      COUNT(city.name) numberOfCities
    FROM
      city,
      countrylanguage
    WHERE city.countrycode = countrylanguage.countrycode
        AND countrylanguage.language = 'English';
```

**Pull request closed without merging**